src/Orleans.Core/Networking/Connection.cs

```diff
@@ -408,5 +400,50 @@ private static EndPoint NormalizeEndpoint(EndPoint endpoint)
 
             return ep;
         }
+
+        /// <summary>
+        /// Handles a message receive failure.
+        /// </summary>
+        /// <returns><see langword="true"/> if the exception should not be caught and <see langword="false"/> if it should be caught.</returns>
+        private bool HandleReceiveMessageFailure(Message message, Exception exception)
+        {
+            this.Log.LogWarning(
+                "Exception reading message {Message} from remote endpoint {Remote} to local endpoint {Local}: {Exception}",
+                message,
+                this.RemoteEndPoint,
+                this.LocalEndPoint,
+                exception);
+
+            // If deserialization completely failed, rethrow the exception so that it can be handled at another level.
+            if (message?.Headers is null)
+            {
+                // Returning false here informs the caller that the exception should not be caught.
+                return false;
+            }
+
+            // The message body was not successfully decoded, but the headers were.
+            MessagingStatisticsGroup.OnRejectedMessage(message);
+
+            if (message.Direction == Message.Directions.Request)
+            {
+                // Send a fast fail to the caller.
+                var response = this.MessageFactory.CreateResponseMessage(message);
+                response.Result = Message.ResponseTypes.Error;
+                response.BodyObject = Response.ExceptionResponse(exception);
+
+                // Send the error response and continue processing the next message.
+                this.MessageCenter.SendMessage(response);
+            }
+            else if (message.Direction == Message.Directions.Response)
+            {
+                // If the message was a response, propagate the exception to the intended recipient.
+                message.Result = Message.ResponseTypes.Error;
```

Comment:
Wouldn't it be better to create a new message here instead of mutating the failed one?

Reply:
We can. In the original version of this I had a `Message.ShallowCopy()` method and used that, however this approach of mutating the message is the common pattern from the existing code so I reverted to doing the same.

I can go with the copy approach instead if you prefer it

Reply:
I'm fine with leaving it as is.